File: src/test/java/com/technews/crawling/JavaReleaseNotesCrawlingTest.java
package com.technews.crawling;

import com.technews.common.util.DateUtils;
import lombok.Builder;
import lombok.extern.slf4j.Slf4j;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
public class JavaReleaseNotesCrawlingTest {

    private static final String ORACLE_BASE_URL = "https://www.oracle.com";
    private static final String JDK_8 = "https://www.oracle.com/java/technologies/javase/8u-relnotes.html";
    private static final String JDK_11 = "https://www.oracle.com/java/technologies/javase/11u-relnotes.html";
    private static final String JDK_17 = "https://www.oracle.com/java/technologies/javase/17u-relnotes.html";
    private static final String JDK_21 = "https://www.oracle.com/java/technologies/javase/21u-relnotes.html";

    @Test
    void test() throws IOException {
        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> JDK_8");
        final List<Release> jdk8ReleaseInfo = getReleaseInfo("jdk8", JDK_8);
        jdk8ReleaseInfo.stream().forEach(System.out::println);

        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> JDK_11");
        final List<Release> jdk11ReleaseInfo = getReleaseInfo("jdk11", JDK_11);
        jdk11ReleaseInfo.stream().forEach(System.out::println);

        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> JDK_17");
        final List<Release> jdk17ReleaseInfo = getReleaseInfo("jdk17", JDK_17);
        jdk17ReleaseInfo.stream().forEach(System.out::println);

        System.out.println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> JDK_21");
        final List<Release> jdk21ReleaseInfo = getReleaseInfo("jdk21", JDK_21);
        jdk21ReleaseInfo.stream().forEach(System.out::println);
    }

    private static List<Release> getReleaseInfo(final String project, final String url) throws IOException {
        if (JDK_8.equals(url)) {
            return getLowerJdk8ReleaseInfo(project, url);
        }

        return getExceededJdk8ReleaseInfo(project, url);
    }

    private static List<Release> getExceededJdk8ReleaseInfo(final String project, final String url) throws IOException {
        List<Release> result = new ArrayList<>();

        final Document doc = Jsoup.connect(url).get();
        final Elements items = doc.select(".obullets");
        for (Element item : items) {
            final Elements liElements = item.select("li");
            result.addAll(getReleaseList(project, liElements));
        }
        return result.stream()
                .filter(Release::isNotEmpty)
                .sorted(Comparator.comparing(Release::version).reversed())
                .collect(Collectors.toList());
    }

    private static List<Release> getLowerJdk8ReleaseInfo(final String project, final String url) throws IOException {
        List<Release> result = new ArrayList<>();

        final Document doc = Jsoup.connect(url).get();
        final Elements items = doc.select(".col-item");
        for (Element item : items) {
            final Elements liElements = item.select("li");
            result.addAll(getReleaseList(project, liElements));
        }
        return result.stream()
                .filter(Release::isNotEmpty)
                .sorted(Comparator.comparing(Release::version).reversed())
                .collect(Collectors.toList());
    }

    private static List<Release> getReleaseList(final String project, final Elements liElements) {
        List<Release> result = new ArrayList<>();
        for (final Element li : liElements) {
            final Release release = getRelease(project, li);
            if (release.isEmpty()) {
                continue;
            }

            result.add(release);
        }
        return result;
    }

    private static Release getRelease(final String project, final Element li) {
        final String liText = li.text();
        if (!liText.contains("GA")) {
            return Release.EMPTY;
        }

        try {
            final String version = liText.split(" \\(")[0];
            final String href = li.select("a").first().attr("href");
            final String gaUrl = ORACLE_BASE_URL + href;
            return Release.builder()
                    .createdDt(LocalDate.now().format(DateUtils.CREATED_FORMATTER))
                    .tags(List.of(project, "java", "release"))
                    .project(project)
                    .version(version)
                    .url(gaUrl)
                    .build();
        } catch (Exception e) {
            log.error("Not found matching version.");
        }

        return Release.EMPTY;
    }

    private record Release(
            String id,
            String project,
            String version,
            String date,
            String url,
            List<String> tags,
            String createdDt
    ) {
        public static final Release EMPTY = new Release(null, null, null, null, null, null, null);

        @Builder
        private Release {
        }

        public boolean isEmpty() {
            return this == EMPTY;
        }

        public boolean isNotEmpty() {
            return !this.isEmpty();
        }
    }
}

File: src/test/java/com/technews/crawling/release/JavaReleaseNotesCrawlingTest.kt
package com.technews.crawling.release

import com.technews.common.util.DateUtils
import io.kotest.core.spec.style.BehaviorSpec
import mu.KotlinLogging
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.time.LocalDate

private val logger = KotlinLogging.logger {}

class JavaReleaseNotesCrawlingTest : BehaviorSpec({
    Given("Java Release 정보 크롤링") {
        val jdks = mapOf(
            "jdk8" to JDK_8,
            "jdk11" to JDK_11,
            "jdk17" to JDK_17,
            "jdk21" to JDK_21,
            "jdk23" to JDK_23,
        )

        jdks.forEach { (project, url) ->
            println(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> $project")
            val releaseInfo = getReleaseInfo(project, url)
            releaseInfo.forEach { println(it) }
        }
    }
}) {
    companion object {
        private const val ORACLE_BASE_URL = "https://www.oracle.com"
        private const val JDK_8 = "https://www.oracle.com/java/technologies/javase/8u-relnotes.html"
        private const val JDK_11 = "https://www.oracle.com/java/technologies/javase/11u-relnotes.html"
        private const val JDK_17 = "https://www.oracle.com/java/technologies/javase/17u-relnotes.html"
        private const val JDK_21 = "https://www.oracle.com/java/technologies/javase/21u-relnotes.html"
        private const val JDK_23 = "https://www.oracle.com/java/technologies/javase/23u-relnotes.html"

        private fun getReleaseInfo(project: String, url: String): List<Release> {
            return if (url == JDK_8) {
                getReleases(project, url, ".col-item")
            } else {
                getReleases(project, url, ".obullets")
            }
        }

        private fun getReleases(project: String, url: String, selector: String): List<Release> {
            return runCatching {
                val doc = Jsoup.connect(url).get()
                doc.select(selector)
                    .flatMap { it.select("li") }
                    .mapNotNull { getRelease(project, it) }
                    .sortedByDescending { it.version }
            }.getOrElse {
                logger.error("Failed to fetch releases for $project: ${it.message}")
                emptyList()
            }
        }

        private fun getRelease(project: String, li: Element): Release? {
            val liText = li.text()
            if (!liText.contains("GA")) return null

            return runCatching {
                val version = liText.split(" \\(".toRegex()).firstOrNull() ?: return null
                val href = li.select("a").firstOrNull()?.attr("href") ?: return null
                val gaUrl = ORACLE_BASE_URL + href
                Release(
                    project = project,
                    version = version,
                    url = gaUrl,
                    tags = listOf(project, "java", "release"),
                    createdDt = LocalDate.now().format(DateUtils.CREATED_FORMATTER),
                )
            }.getOrElse {
                println("Failed to parse release for $project: ${it.message}")
                null
            }
        }

        private data class Release(
            val project: String,
            val version: String,
            val url: String,
            val tags: List<String>,
            val createdDt: String,
        )
    }
}
